samples/wifi/promiscuous/src/wifi_connection.c

```diff
@@ -0,0 +1,233 @@
+	}
+	return 0;
+}
+
+static void handle_wifi_connect_result(struct net_mgmt_event_callback *cb)
+{
+	const struct wifi_status *status =
+		(const struct wifi_status *) cb->info;
+
+	if (context.connected) {
+		return;
+	}
+
+	if (status->status) {
+		LOG_ERR("Connection failed (%d)", status->status);
+	} else {
+		LOG_INF("Connected");
+		context.connected = true;
+	}
+
+	context.connect_result = status->conn_status;
+}
+
+static void handle_wifi_disconnect_result(struct net_mgmt_event_callback *cb)
+{
+	const struct wifi_status *status =
+		(const struct wifi_status *) cb->info;
+
+	if (!context.connected) {
+		return;
+	}
+
+	if (context.disconnect_requested) {
+		LOG_INF("Disconnection request %s (%d)",
+			 status->status ? "failed" : "done",
+					status->status);
+		context.disconnect_requested = false;
+	} else {
+		LOG_INF("Received Disconnected");
+		context.connected = false;
+	}
+
+	cmd_wifi_status();
+}
+
+static void wifi_mgmt_event_handler(struct net_mgmt_event_callback *cb,
+				     uint32_t mgmt_event, struct net_if *iface)
+{
+	switch (mgmt_event) {
+	case NET_EVENT_WIFI_CONNECT_RESULT:
+		handle_wifi_connect_result(cb);
+		break;
+	case NET_EVENT_WIFI_DISCONNECT_RESULT:
+		handle_wifi_disconnect_result(cb);
+		break;
+	default:
+		break;
+	}
+}
+
+static void print_dhcp_ip(struct net_mgmt_event_callback *cb)
+{
+	/* Get DHCP info from struct net_if_dhcpv4 and print */
+	const struct net_if_dhcpv4 *dhcpv4 = cb->info;
+	const struct in_addr *addr = &dhcpv4->requested_ip;
+	char dhcp_info[128];
+
+	net_addr_ntop(AF_INET, addr, dhcp_info, sizeof(dhcp_info));
+
+	LOG_INF("DHCP IP address: %s", dhcp_info);
+	k_sem_give(&wait_for_dhcp);
+}
+
+static void net_mgmt_event_handler(struct net_mgmt_event_callback *cb,
+				    uint32_t mgmt_event, struct net_if *iface)
+{
+	switch (mgmt_event) {
+	case NET_EVENT_IPV4_DHCP_BOUND:
+		print_dhcp_ip(cb);
+		break;
+	default:
+		break;
+	}
+}
+
+static int wifi_connect(void)
+{
+	struct net_if *iface = net_if_get_first_wifi();
+
+	if (!iface) {
+		LOG_ERR("Failed to get Wi-Fi iface");
+		return -1;
+	}
+
+	context.connected = false;
+	context.connect_result = WIFI_STATUS_CONN_FAIL;
+
+	if (net_mgmt(NET_REQUEST_WIFI_CONNECT_STORED, iface, NULL, 0)) {
+		LOG_ERR("Connection request failed");
+		return -ENOEXEC;
+	}
+
+	LOG_INF("Connection requested");
+
+	return 0;
+}
+
+int try_wifi_connect(void)
+{
+	memset(&context, 0, sizeof(context));
+
+	net_mgmt_init_event_callback(&wifi_shell_mgmt_cb,
+				     wifi_mgmt_event_handler,
+				     WIFI_SHELL_MGMT_EVENTS);
+
+	net_mgmt_add_event_callback(&wifi_shell_mgmt_cb);
+
+	net_mgmt_init_event_callback(&net_shell_mgmt_cb,
+				     net_mgmt_event_handler,
+				     NET_EVENT_IPV4_DHCP_BOUND);
+
+	net_mgmt_add_event_callback(&net_shell_mgmt_cb);
+
+	k_sleep(K_SECONDS(1));
+
+	LOG_INF("Static IP address (overridable): %s/%s -> %s",
+		CONFIG_NET_CONFIG_MY_IPV4_ADDR,
+		CONFIG_NET_CONFIG_MY_IPV4_NETMASK,
+		CONFIG_NET_CONFIG_MY_IPV4_GW);
+
+	while (1) {
```

Comment:
can we convert this to a non-blocking call where we issue a connect and wait for result/timeout? IIRC, timeout should also generate a failure event.